### ISSUE TITLE
Design: Modelling

### ISSUE THREAD

**Issue by abrichr** [OpenAdaptAI/OpenAdapt#72]
There is some core modelling work that needs proper attention.

We need folks who are knowledgeable/experienced/motivated to spend an hour or two thinking about a few ways to model this, and what an MVP technical implementation looks like.

I suspect it may be https://github.com/MLDSAI/PuterBot/issues/12 and/or https://github.com/MLDSAI/PuterBot/issues/63, but there may be a more direct and simpler way.

Ultimately we would like to publish a paper on the results.

Useful starting points:
- how does https://www.adept.ai/blog/act-1 work exactly?
- what are limitations of that approach?

In particular, we want to extend our models to include:
- screen shots (vision)
- structured window data
- structured action data
- natural language task descriptions
- API documentation

Related: https://github.com/MLDSAI/OpenAdapt/issues/252

**Comment by angelala3252** [comment]
Starting point:
ACT-1 is a Chrome extension that takes a user prompt for a task to perform and then carries out that task. Since it is not an open-source project, I don't know 100% how it works, but going off the blog post and what Richard has explained about it, it likely follows these steps:

1. takes in natural language input from the user describing the task
2. processes the input using transformers to understand what the task is
3. renders the code of the webpage and identifies the UI elements it needs to interact with to accomplish its task
4. completes the task

Limitations:
- only works on Chrome browsers as it uses only the HTML/CSS code on the webpage, and doesn't use image recognition

Our model:
For now, based on [this comment](https://github.com/MLDSAI/OpenAdapt/pull/98#issuecomment-1537468937) and the work I've done so far on [issue #12 ](https://github.com/MLDSAI/OpenAdapt/issues/12), I think an MVP of OpenAdapt would look something like this:

1. take in natural language input from the user describing the task
2. give the input to MiniGPT-4 (or another LLM) to process the input using transformers to understand what the task is and what next steps need to be done
3. check if there are recordings of that step which we can then playback using some strategy
4. allow for prompting the user to take over if a recording does not exist 
5. playback each step until the task is completed

I'm a little confused about whether there will be user input for step 1 or not, because it seems that in the ChatGPT plugin demo, we are providing a task description, but in the readme, it says that "Instead of requiring the user to prompt the model directly, we prompt it behind the scenes by observing the user's activities"

I'm open to any feedback or suggestions about possible designs!

**Comment by 0dm** [comment]
the ultimate goal is to prompt behind the scenes rather than requiring the user to do it

**Comment by FFFiend** [comment]
I don't think ACT-1 is a browser plugin, they simply hooked it up to Chrome to teach it how it works and prompted it based on its training.

**Comment by abrichr** [comment]
Formalism (work in progress):

R: Recording = { A_t, S_t, W_t }
A_M(t): Action in Modality taken at time t
M: Modality = { Keyboard, Mouse }
S(t): Screenshot taken immediately at time t (most recently before Action)
W(t): Window state recorded at time t (most recently before Action)
T_L: Task at Level
L: Level of abstraction (https://github.com/MLDSAI/OpenAdapt/issues/185)
P: [Policy](https://en.wikipedia.org/wiki/Reinforcement_learning#Policy) (a.k.a Strategy)
R: Role - the person responsible for completing a task
C: Context - details that surround the task and what makes it relevant to the role
W: Workflow - the sequence of tasks required to deliver the expected outcome
G: [Signal](https://github.com/MLDSAI/OpenAdapt/pull/160) measuring expected outcome

**Comment by flyguy712** [comment]
This is a simplified example just meant to elaborate on the definitions...

- Role: Carwash attendant
- Task: Spray bugs off the car before it rolls through the automated car wash
- Context:
  - Some cars have a lot of bugs and thus require focused of spraying
  - Some cars are bigger than others, and thus require distributed spraying
  - There are limitations to the spraying time that management allows for each car to save costs
  - Some employees are left handed and have to move the hose to their right hand to trigger the automated wash via the wall computer for each vehicle
  - Attendants must use different spraying patters to cover the surface area for each car to accomplish the task in the allotted spraying time
- Workflow:
  - Small car, many bugs:
    - Car pulls up to the car wash entrance
    - Attendant waives car in with a hand motion
    - As car crosses garage threshold, attendant starts spraying in a side to side motion just focusing on the concentration of maximum bug build up on the bumper
    - After 5 seconds employee stops spraying
    - Employee activates automated car wash routine via wall computer
  - Big car, few bugs:
    - Car pulls up to the car wash entrance
    - Attendant waives car in with a hand motion
    - As car crosses garage threshold, attendant starts spraying in a side to side motion for 2 passes, then switches to a top down motion for 2 passes, and so on, to provide maximum area coverage of the entire front of the car
    - After 5 seconds employee stops spraying
    - Employee activates automated car wash routine via wall computer

In the context of [Jobs To Be Done](https://hbr.org/2016/09/know-your-customers-jobs-to-be-done) (which can be summarized as 'the progress that the customer is trying to make in a given circumstance'), you can think of the definitions as follows:

- Role : customer
- Task : progress
- Context : circumstance
- Workflow : workflow

Another more relevant example...

- Role: Travel agent
- Task: Book a flight
- Context:
  - There are 4 people traveling
  - 2 travelers require first class tickets
  - 1 traveler prefers aile seats
  - 1 traveler prefers window seats
  - The flight is to Barcelona
  - There can't be more than 1 layover
  - Total ticket prices must cost less than $10,000
- Workflow:
  - Open Chrome
  - Navigate to Google Flights
  - Set the departure city
  - Set the arrival city
  - Select the departure date
  - Select the return date
  - Click "Search"
  - Update number of travelers by clicking on person icon and selecting "4"
  - Update price filter to $10,000
  - Update stops to "1 stops or fewer"
  - ...
  
**TL;DR - the context is what dictates the workflow for a given task.**

**Comment by abrichr** [comment]
Work in progress: https://app.diagrams.net/#G1_fYoFncuI0TKghKWiMvP13hHEnIqAHI_

**Comment by abrichr** [comment]
https://github.com/OpenAdaptAI/OpenAdapt/discussions/552
